src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs

```diff
@@ -34,270 +28,256 @@
-                        // wait context. That means the command system won't attempt to show its own wait dialog 
-                        // and also will take it into consideration when measuring command handling duration.
-                        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-
-                        if (!notificationService.ConfirmMessageBox(
-                                EditorFeaturesResources.Extract_method_encountered_the_following_issues + Environment.NewLine + Environment.NewLine +
-                                string.Join(Environment.NewLine, result.Reasons) + Environment.NewLine + Environment.NewLine +
-                                EditorFeaturesResources.We_can_fix_the_error_by_not_making_struct_out_ref_parameter_s_Do_you_want_to_proceed,
-                                title: EditorFeaturesResources.Extract_Method,
-                                severity: NotificationSeverity.Error))
-                        {
-                            // We handled the command, displayed a notification and did not produce code.
-                            return;
-                        }
-
-                        await TaskScheduler.Default;
-                    }
-
-                    // reset result
-                    result = newResult;
-                }
-                else if (await TryNotifyFailureToUserAsync(document, result, cancellationToken).ConfigureAwait(false))
-                {
-                    // We handled the command, displayed a notification and did not produce code.
-                    return;
-                }
-            }
+        var cancellationToken = waitContext.UserCancellationToken;
 
-            var cleanupOptions = await document.GetCodeCleanupOptionsAsync(_globalOptions, cancellationToken).ConfigureAwait(false);
-            var (formattedDocument, methodNameAtInvocation) = await result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).ConfigureAwait(false);
-            var changes = await formattedDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+        var document = await textBuffer.CurrentSnapshot.GetFullyLoadedOpenDocumentInCurrentContextWithChangesAsync(waitContext).ConfigureAwait(false);
+        if (document is null)
+            return;
 
-            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+        var options = await document.GetExtractMethodGenerationOptionsAsync(_globalOptions, cancellationToken).ConfigureAwait(false);
+        var result = await ExtractMethodService.ExtractMethodAsync(
+            document, span, localFunction: false, options, cancellationToken).ConfigureAwait(false);
+        Contract.ThrowIfNull(result);
 
-            ApplyChange_OnUIThread(textBuffer, changes, waitContext);
+        result = await NotifyUserIfNecessaryAsync(
```

Comment:
extracted this out to make the code flow and individual cases easier to understand.